### PR TITLE
Use similar+copy! instead of copy_oftype for AbstractMatrix

### LINKED ISSUE #316: `a \ b::Diagonal`  throws an error
On 0.4.3 and master:

```
julia> rand(5,5)\Diagonal(ones(5))
ERROR: MethodError: `A_ldiv_B!` has no method matching 
A_ldiv_B!(::Base.LinAlg.LU{Float64,Array{Float64,2}}, ::Diagonal{Float64})
```

I am not sure if a specialized method would make sense but in any case falling back on `a\full(b)` would  seem reasonable.

### LINKED ISSUE #15401: `a \ b::Diagonal`  throws an error
On 0.4.3 and master:
~~~
julia> rand(5,5)\Diagonal(ones(5))
ERROR: MethodError: `A_ldiv_B!` has no method matching 
A_ldiv_B!(::Base.LinAlg.LU{Float64,Array{Float64,2}}, ::Diagonal{Float64})
~~~
I am not sure if a specialized method would make sense but in any case falling back on `a\full(b)` would  seem reasonable.

### LINKED ISSUE #16201: Infinitely growing compilation recursion
On 0.4 I get
```jl
julia> UpperTriangular(randn(3,3))\sub(randn(3), [1,2,3])
ERROR: StackOverflowError:
 in \ at linalg/generic.jl:323
 in \ at linalg/generic.jl:326 (repeats 74845 times)
```
but on `0.5.0-dev+3878 (2016-05-04 17:25 UTC)` Julia just hangs and I cannot break out of recursion so I have to kill Julia.

Update: This is on OS X.

### PR DESCRIPTION
**Pull request by andreasnoack**
Throw error for `LowerTriangular(::UpperTriagular)` and vice versa to avoid infinite compilation recursion.

Use `similar`+`copy!` instead of `copy_oftype` for `AbstractMatrix` arguments to make sure that the copy is mutable. Fixes JuliaLang/LinearAlgebra.jl#316

This also kind of fixes https://github.com/JuliaLang/julia/issues/16201 since it doesn't hang anymore.

### CODE REVIEW

File: base/linalg/triangular.jl [hunk 1]
```diff
@@ -39,6 +39,9 @@ for t in (:LowerTriangular, :UnitLowerTriangular, :UpperTriangular,
     end
 end
 
+LowerTriangular(U::UpperTriangular) = throw(ArgumentError("this shouldn't happen. Please submit a bug report"))
+UpperTriangular(U::LowerTriangular) = throw(ArgumentError("this shouldn't happen. Please submit a bug report"))
```
Comment (tkelman) [inline base/linalg/triangular.jl]: I don't think this is quite right, a user could easily call the constructors to try to do this. Maybe something like "Cannot create an UpperTriangular matrix from a LowerTriangular input"?

File: base/linalg/cholesky.jl [hunk 2]
```diff
@@ -86,7 +86,9 @@ and return the UpperTriangular matrix `U` such that `A = U'U`.
 """
 function chol{T}(A::AbstractMatrix{T})
     S = promote_type(typeof(chol(one(T))), Float32)
-    chol!(copy_oftype(A, S))
+    AA = similar(A, S, size(A))
+    copy!(AA, A)
```
Comment (tkelman) [inline base/linalg/cholesky.jl]: Why isn't `copy_oftype` written using `similar` to just do this?

Comment (andreasnoack) [inline base/linalg/cholesky.jl]: I noticed that we use `copy_oftype` in places where we don't want the fully mutable matrix, e.g. https://github.com/JuliaLang/julia/blob/master/base/linalg/tridiag.jl#L128. I thought about `mutablecopy` but decided that it was almost as simple just to write it out.

File: base/linalg/triangular.jl (hunk 1 above)
Comment (andreasnoack) [inline base/linalg/triangular.jl]: Yes. That is probably better. I'll update the error message.

**Pull request merged**